File: src/handlers.py
import asyncio
import base64
import io
import logging
import time
import traceback
from dataclasses import asdict
from typing import TypedDict

from aiogram import Bot, F, Router
from aiogram.types import Message
from aiogram.enums import ParseMode

from .config import BOT_TOKEN, CHAT_IDS, SUMMARY_INTERVAL, SUMMARY_INTERVALS, WIDE_LOG_DIR
from .db import (
    get_last_summary_id,
    get_messages,
    get_sticker,
    save_message,
    save_sticker,
    save_summary,
    set_last_summary_id,
)
from .llm import (
    describe_image, describe_sticker, describe_video_note, describe_voice,
    generate_summary, get_model_short_name,
)
from .utils import append_wide_log, fix_html, get_attachment_info, get_message_text, get_sender_name, log_context, log_warning, logged, timed

logger = logging.getLogger(__name__)

bot = Bot(token=BOT_TOKEN)
router = Router()


# Lock для предотвращения одновременных генераций (per-chat)
summary_locks: dict[int, asyncio.Lock] = {}
generating_chats: set[int] = set()


async def send_summary(chat_id: int, summary: str, model: str):
    """Отправляет саммари в чат."""
    summary = summary[:3000]
    summary = fix_html(summary)
    
    model_short = get_model_short_name(model)
    full_message = f"#concisely\n{summary}\n\n{model_short}"
    
    try:
        await bot.send_message(chat_id, full_message, parse_mode=ParseMode.HTML)
    except Exception as e:
        log_warning(f"html_fallback: {e}")
        await bot.send_message(chat_id, full_message)


class SummaryInfo(TypedDict, total=False):
    """Результат maybe_generate_summary — данные для лога и БД."""
    # Статус
    attempted: bool
    sent: bool
    reason: str
    error: str
    # Контекст
    last_summary_id: int | None
    messages_since_last: int
    interval: int
    messages_count: int
    timing_ms: float
    # Данные саммари (если sent)
    chat_id: int
    from_message_id: int
    to_message_id: int
    text: str
    model: str
    input_tokens: int | None
    output_tokens: int | None
    cost: float | None


async def _generate_and_send_summary(chat_id: int, from_id: int, to_id: int) -> SummaryInfo:
    """Генерирует, отправляет и сохраняет саммари. Возвращает данные для лога и БД."""
    messages = await get_messages(chat_id, from_id, to_id)
    if not messages:
        return {"messages_count": 0, "reason": "no_messages"}
    
    result = await generate_summary(messages)
    await send_summary(chat_id, result.text, result.model)
    await set_last_summary_id(chat_id, to_id)
    
    return {
        "chat_id": chat_id,
        "from_message_id": from_id,
        "to_message_id": to_id,
        "messages_count": len(messages),
        "sent": True,
        **asdict(result),
    }


@logged("summary")
@timed("summary")
async def maybe_generate_summary(current_message_id: int, chat_id: int) -> SummaryInfo:
    """Проверяет, нужно ли генерировать саммари, и генерирует если нужно."""
    info: SummaryInfo = {"attempted": False, "sent": False}
    
    if chat_id in generating_chats:
        info["reason"] = "already_generating"
        return info
    
    if chat_id not in summary_locks:
        summary_locks[chat_id] = asyncio.Lock()
    
    async with summary_locks[chat_id]:
        if chat_id in generating_chats:
            info["reason"] = "already_generating"
            return info
        
        last_summary_id = await get_last_summary_id(chat_id)
        info["last_summary_id"] = last_summary_id
        
        if last_summary_id is None:
            await set_last_summary_id(chat_id, current_message_id)
            info["reason"] = "first_run"
            return info
        
        interval = SUMMARY_INTERVALS.get(chat_id, SUMMARY_INTERVAL)
        if current_message_id - last_summary_id < interval:
            info["reason"] = "interval_not_reached"
            info["messages_since_last"] = current_message_id - last_summary_id
            info["interval"] = interval
            return info
        
        generating_chats.add(chat_id)
    
    try:
        info["attempted"] = True
        
        gen_result = await _generate_and_send_summary(chat_id, last_summary_id, current_message_id)
        info.update(gen_result)
        
    except Exception as e:
        info["reason"] = "error"
        info["error"] = str(e)
    finally:
        generating_chats.discard(chat_id)
    
    return info


async def _download_file_bytes(file_id: str) -> bytes:
    """Скачивает файл из Telegram и возвращает сырые байты."""
    file = await bot.get_file(file_id)
    buffer = io.BytesIO()
    await bot.download_file(file.file_path, buffer)
    buffer.seek(0)
    return buffer.read()


async def _download_file_base64(file_id: str) -> str:
    """Скачивает файл из Telegram и возвращает base64."""
    raw = await _download_file_bytes(file_id)
    return base64.b64encode(raw).decode('utf-8')


class DescribeInfo(TypedDict):
    description: str
    cost: float | None


@logged
@timed
async def describe_attachment(message: Message, attachment: dict) -> DescribeInfo | None:
    """Получает описание вложения от vision-модели."""
    att_type = attachment["type"]
    
    try:
        if att_type == "photo" and message.photo:
            b64 = await _download_file_base64(message.photo[-1].file_id)
            result = await describe_image(b64)
        
        elif att_type == "sticker" and message.sticker:
            sticker = message.sticker
            cached = await get_sticker(sticker.file_unique_id)
            if cached is not None:
                return {"description": cached, "cost": None}
            # Для анимированных/видео стикеров используем thumbnail
            if sticker.is_animated or sticker.is_video:
                if not sticker.thumbnail:
                    log_warning(f"sticker {sticker.file_unique_id}: нет thumbnail")
                    return None
                file_id = sticker.thumbnail.file_id
            else:
                file_id = sticker.file_id
            b64 = await _download_file_base64(file_id)
            result = await describe_sticker(b64)
            await save_sticker(sticker.file_unique_id, result.text)
        
        elif att_type == "voice" and message.voice:
            raw = await _download_file_bytes(message.voice.file_id)
            result = await describe_voice(raw)
        
        elif att_type == "video_note" and message.video_note:
            b64 = await _download_file_base64(message.video_note.file_id)
            result = await describe_video_note(b64)
        
        else:
            return None
    except Exception as e:
        log_warning(f"describe_{att_type}: {e}")
        return None
    
    return {"description": result.text, "cost": result.cost}


@router.message(F.chat.id.in_(CHAT_IDS))
async def handle_message(message: Message):
    """Обрабатывает все сообщения из отслеживаемых чатов."""
    context: dict = {"timings": {}}
    token = log_context.set(context)
    
    start = time.perf_counter()
    try:
        attachment = get_attachment_info(message)
        raw_message = message.model_dump(
            mode="json", exclude_none=True, exclude_unset=True, exclude_defaults=True,
        )
        context["request_id"] = f"{message.chat.id}:{message.message_id}"
        context["message"] = raw_message
        
        if attachment:
            describe = await describe_attachment(message, attachment)
            if describe:
                attachment.update(describe)
        
        await save_message({
            "chat_id": message.chat.id,
            "message_id": message.message_id,
            "sender_id": message.from_user.id if message.from_user else None,
            "sender_name": get_sender_name(message),
            "text": get_message_text(message),
            "reply_to_message_id": message.reply_to_message.message_id if message.reply_to_message else None,
            "timestamp": message.date.isoformat() if message.date else None,
            "raw": raw_message,
            "attachment": attachment,
        })
        
        summary_info = await maybe_generate_summary(message.message_id, message.chat.id)
        if summary_info.get("sent"):
            await save_summary(summary_info)
        
    except Exception as e:
        context["error"] = str(e)
        context["error_traceback"] = traceback.format_exc()
    finally:
        context["timings"]["total"] = round((time.perf_counter() - start) * 1000, 2)
        log_context.reset(token)
        try:
            append_wide_log(context, WIDE_LOG_DIR)
        except Exception as e:
            logger.warning(f"Не удалось записать wide log: {e}")

File: src-ts/handlers.ts
import type { Chat, Message } from 'grammy/types';
import { bot } from './bot.ts';
import { CHATS, CHAT_IDS, BOT_TOKEN, WIDE_LOG_DIR } from './config.ts';
import {
  getLastSummaryId, setLastSummaryId, saveMessage, getMessages,
  getSticker, saveSticker, saveSummary, cleanupOldMessages,
  type SummaryData, type CleanupResult,
} from './db.ts';
import {
  generateSummary, describeImage, describeSticker, describeVoice, describeVideoNote,
  getModelShortName, type MessageData,
} from './llm.ts';
import {
  logStorage, logWarning, timed, logged, timedAndLogged,
  fixHtml, getMessageText, getAttachmentInfo, getSenderName, getForwardSenderName,
  appendWideLog, type LogContext,
} from './utils.ts';

const summaryLocks = new Map<number, Promise<void>>();
const generatingChats = new Set<number>();

async function sendSummary(chatId: number, summary: string, model: string, threadId?: number): Promise<void> {
  const text = fixHtml(summary.slice(0, 3000));
  const modelShort = getModelShortName(model);
  const fullMessage = `#concisely\n${text}\n\n${modelShort}`;

  try {
    await bot.api.sendMessage(chatId, fullMessage, { parse_mode: 'HTML', message_thread_id: threadId });
  } catch (e) {
    logWarning(`html_fallback: ${e}`);
    await bot.api.sendMessage(chatId, fullMessage, { message_thread_id: threadId });
  }
}

async function keepTyping(chatId: number, signal: AbortSignal, threadId?: number): Promise<void> {
  while (!signal.aborted) {
    await bot.api.sendChatAction(chatId, 'typing', { message_thread_id: threadId });
    await new Promise<void>((resolve, reject) => {
      const t = setTimeout(resolve, 4000);
      signal.addEventListener('abort', () => { clearTimeout(t); reject(new Error('aborted')); }, { once: true });
    }).catch(() => {});
    if (signal.aborted) break;
  }
}

const generateAndSendSummary = timed('summary', async (chatId: number, fromId: number, toId: number, threadId?: number): Promise<SummaryData | null> => {
  const messages = await getMessages(chatId, fromId, toId) as MessageData[];
  if (!messages.length) return null;

  const ac = new AbortController();
  const typingLoop = keepTyping(chatId, ac.signal, threadId);
  let result: Awaited<ReturnType<typeof generateSummary>>;
  try {
    result = await generateSummary(messages);
  } finally {
    ac.abort();
    await typingLoop;
  }

  await sendSummary(chatId, result.text, result.model, threadId);
  await setLastSummaryId(chatId, toId);

  return {
    chat_id: chatId,
    from_message_id: fromId,
    to_message_id: toId,
    text: result.text,
    model: result.model,
    input_tokens: result.input_tokens,
    output_tokens: result.output_tokens,
    cost: result.cost,
  };
});

interface SummaryInfo {
  attempted: boolean;
  sent: boolean;
  reason?: string;
  error?: string;
  last_summary_id?: number | null;
  messages_since_last?: number;
  interval?: number;
  data?: SummaryData;
  timing_ms?: number;
  cleanup?: CleanupResult;
}

const maybeGenerateSummary = logged('summary', async (currentMessageId: number, chatId: number, threadId?: number): Promise<SummaryInfo> => {
  const info: SummaryInfo = { attempted: false, sent: false };

  if (generatingChats.has(chatId)) {
    info.reason = 'already_generating';
    return info;
  }

  // Serialize per-chat
  const prevLock = summaryLocks.get(chatId) ?? Promise.resolve();
  let resolveLock!: () => void;
  const lock = new Promise<void>(r => { resolveLock = r; });
  summaryLocks.set(chatId, prevLock.then(() => lock));

  await prevLock;

  if (generatingChats.has(chatId)) {
    resolveLock();
    info.reason = 'already_generating';
    return info;
  }

  const lastSummaryId = await getLastSummaryId(chatId);
  info.last_summary_id = lastSummaryId;

  if (lastSummaryId === null) {
    await setLastSummaryId(chatId, currentMessageId);
    resolveLock();
    info.reason = 'first_run';
    return info;
  }

  const chatConfig = CHATS.get(chatId)!;
  if (currentMessageId - lastSummaryId < chatConfig.interval) {
    resolveLock();
    info.reason = 'interval_not_reached';
    info.messages_since_last = currentMessageId - lastSummaryId;
    info.interval = chatConfig.interval;
    return info;
  }

  generatingChats.add(chatId);
  resolveLock();

  try {
    info.attempted = true;
    const data = await generateAndSendSummary(chatId, lastSummaryId, currentMessageId, threadId);
    if (data) {
      info.sent = true;
      info.data = data;
      info.cleanup = await cleanupOldMessages(chatId, data.to_message_id, chatConfig.interval);
    } else {
      info.reason = 'no_messages';
    }
  } catch (e) {
    info.reason = 'error';
    info.error = String(e);
  } finally {
    generatingChats.delete(chatId);
  }

  return info;
});

async function downloadFileBytes(fileId: string): Promise<Uint8Array> {
  const file = await bot.api.getFile(fileId);
  const url = `https://api.telegram.org/file/bot${BOT_TOKEN}/${file.file_path}`;
  const res = await fetch(url);
  return new Uint8Array(await res.arrayBuffer());
}

async function downloadFileBase64(fileId: string): Promise<string> {
  const bytes = await downloadFileBytes(fileId);
  return Buffer.from(bytes).toString('base64');
}

interface DescribeInfo {
  description: string;
  cost: number | null;
}

const describeAttachment = timedAndLogged('describe_attachment', async (message: Message, attachment: Record<string, unknown>): Promise<DescribeInfo | null> => {
  const attType = attachment.type as string;
  try {
    if (attType === 'photo' && message.photo) {
      const b64 = await downloadFileBase64(message.photo.at(-1)!.file_id);
      const result = await describeImage(b64);
      return { description: result.text, cost: result.cost };
    }

    if (attType === 'sticker' && message.sticker) {
      const sticker = message.sticker;
      const cached = await getSticker(sticker.file_unique_id);
      if (cached !== null) return { description: cached, cost: null };

      const fileId = (sticker.is_animated || sticker.is_video)
        ? sticker.thumbnail?.file_id
        : sticker.file_id;

      if (!fileId) {
        logWarning(`sticker ${sticker.file_unique_id}: нет thumbnail`);
        return null;
      }
      const b64 = await downloadFileBase64(fileId);
      const result = await describeSticker(b64);
      await saveSticker(sticker.file_unique_id, result.text);
      return { description: result.text, cost: result.cost };
    }

    if (attType === 'voice' && message.voice) {
      const raw = await downloadFileBytes(message.voice.file_id);
      const result = await describeVoice(raw);
      return { description: result.text, cost: result.cost };
    }

    if (attType === 'video_note' && message.video_note) {
      const b64 = await downloadFileBase64(message.video_note.file_id);
      const result = await describeVideoNote(b64);
      return { description: result.text, cost: result.cost };
    }
  } catch (e) {
    logWarning(`describe_${attType}: ${e}`);
  }
  return null;
});

function chatShiftId(chatId: number): string {
  return String(chatId).replace('-100', '');
}

function messageLink(chat: Chat, messageId: number, threadId?: number): string {
  const username = chat.type !== 'group' ? chat.username : undefined;
  const base = username ? `https://t.me/${username}` : `https://t.me/c/${chatShiftId(chat.id)}`;
  return threadId ? `${base}/${threadId}/${messageId}` : `${base}/${messageId}`;
}

async function sendTranscription(chat: Chat, messageId: number, text: string, threadId?: number): Promise<void> {
  const link = messageLink(chat, messageId, threadId);
  const html = `<blockquote expandable><a href="${link}">↑</a> ${fixHtml(text)}</blockquote>`;
  try {
    await bot.api.sendMessage(chat.id, html, { parse_mode: 'HTML', message_thread_id: threadId });
  } catch (e) {
    logWarning(`send_transcription: ${e}`);
  }
}

export function registerHandlers(): void {
  bot.on('message', async (ctx) => {
    const message = ctx.message;
    if (!CHAT_IDS.has(message.chat.id)) return;

    const context: LogContext = { timings: {} };
    await logStorage.run(context, async () => {
      const start = performance.now();
      try {
        const attachment = getAttachmentInfo(message);
        context.request_id = `${message.chat.id}:${message.message_id}`;
        context.message = message;

        const chatConfig = CHATS.get(message.chat.id)!;

        if (attachment) {
          const describe = await describeAttachment(message, attachment);
          if (describe) {
            Object.assign(attachment, describe);
            if ((attachment.type === 'voice' || attachment.type === 'video_note') && chatConfig.transcribe) {
              await sendTranscription(message.chat, message.message_id, describe.description, message.message_thread_id);
            }
          }
        }

        const messageData = {
          chat_id: message.chat.id,
          message_id: message.message_id,
          sender_name: getSenderName(message),
          text: getMessageText(message),
          reply_to_message_id: message.reply_to_message?.message_id ?? null,
          forward_sender_name: getForwardSenderName(message),
          raw: message,
          attachment: attachment ?? null,
        };
        await saveMessage(messageData);

        const summaryInfo = await maybeGenerateSummary(message.message_id, message.chat.id, chatConfig.summary_topic_id);
        if (summaryInfo.sent && summaryInfo.data) {
          await saveSummary(summaryInfo.data);
        }
      } catch (e) {
        context.error = String(e);
        context.error_stack = e instanceof Error ? e.stack : undefined;
      } finally {
        context.timings.total = Math.round((performance.now() - start) * 10) / 10;
      }
    });

    appendWideLog(context, WIDE_LOG_DIR);
  });
}
